### PR TITLE
RFC: Keras SavedModel saving/loading

### PR DESCRIPTION
**Pull request by k-w-w**
**Feedback period open until 2019-05-27**

# Keras SavedModel saving/loading 

 | Status        | Accepted                                             |
:-------------- |:---------------------------------------------------- |
| **Author(s)** | Kathy Wu (kathywu@google.com)                        |
| **Sponsor**   | Karmel (karmel@google.com)                           |
| **Updated**   | 2019-05-30                                           |

 ## Objective

 Add ability to (1) save a [TensorFlow SavedModel](https://www.tensorflow.org/alpha/guide/saved_model) from a Keras model and (2) load a Keras model from a SavedModel. This is a  standalone serialization format, which allows models with custom objects to be saved/loaded without having the original code (currently required by the HDF5 and JSON formats). 

 **Goals:**

 - Add SavedModel format for saving Keras models. 
- Support comprehensive model serialization/deserialization 
  - Define model serialization coverage (what parts of the model are saved to the SavedModel) 
  - Address serialization pattern that is reusable across different frameworks (TensorFlow Hub, tf.module, etc.) 
  - Define SavedModel deserialization (properties of the reconstructed model) 

 **Non-goals:**

 - Serializing custom training loops. Only the standard `model.compile` to  `model.fit` use case is covered.
- Changing existing serialization formats

### CODE REVIEW

**Comment by frreiss** [comment]
What will happen to `tf.keras`'s HDF5 and JSON support after this change goes in? Will users of `tf.keras` need to use SavedModel instead of HDF5 or JSON to make use of the full capabilities of the TensorFlow version of Keras?

**Comment by k-w-w** [comment]
@frreiss `model.save` will have a `save_format` argument, which can be set to continue saving in the HDF5 format.

File: rfcs/20190509-keras-saved-model.md [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+  - Define model serialization coverage (what parts of the model are saved 
+    to the SavedModel) 
+  - Address serialization pattern that is reusable across different
+    frameworks (TensorFlow Hub, tf.module, etc.) 
+  - Define SavedModel deserialization (properties of the reconstructed 
+    model) 
+
+**Non-goals:**
+
+- Serializing custom training loops. Only the standard `model.compile` to
```
Comment (alextp) [inline rfcs/20190509-keras-saved-model.md]: To confirm, custom training loops will not be serialized by default, but it will still be possible to load a keras saved model and write a custom training loop for it, right?

Also we should probably say that if you want to serialize your custom training loop you can wrap it in tf.function and export that function as part of the savedmodel, right?

File: rfcs/20190509-keras-saved-model.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+When reconstructing a Keras model, the saved attributes are remapped to the original names. The exception is the call function, which uses `call_and_return_conditional_losses` instead of `__call__`.
+
+## Questions and Discussion Topics
+
+1. Do we all agree on the API changes?
+2. Are there other aspects of the model that should be serialized/deserialized?
+3. Should `tf.saved_model.load` return a generic object or Keras model?
+4. Syncing common endpoints with tf.module
+
```
Comment (alextp) [inline rfcs/20190509-keras-saved-model.md]: Can you document the backward and forward compatibility guarantees of this serialization format? I assume it'll be fully backwards compatible.

Re forward compatibility, though, what happens if one tries to load a model exported from tf version X into a tf version Y < X? Is the user guaranteed to get all the features that Y supports?

File: rfcs/20190509-keras-saved-model.md (hunk 1 above)
Comment (k-w-w) [reply rfcs/20190509-keras-saved-model.md]: Yes, the loaded keras model will behave identically to the original model. And yeah, attaching the training loop to the model will work. Added a comment to the doc.

File: rfcs/20190509-keras-saved-model.md (hunk 2 above)
Comment (k-w-w) [reply rfcs/20190509-keras-saved-model.md]: Forward compatibility seems feasible. I added a compatibility section promising both backward and forward compatibility.

File: rfcs/20190509-keras-saved-model.md [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+The following checkpointable objects and functions are attached to the saved Keras model:
+
+- `variables`: List of all variables in this layer and submodules.
+- `trainable_variables`: List of all trainable variables in this layer and submodules.
+- `non_trainable_variables`: List of all non-trainable variables in this layer and submodules.
+- `regularization_losses`: List of unconditional loss functions in this layer 
+  and submodules. Each function takes no arguments, and returns a scalar 
+  tensor. 
+- `submodules`: Flat list of all sublayers (does not include metrics, even though Metric subclasses Layer).
```
Comment (fchollet) [inline rfcs/20190509-keras-saved-model.md]: We don't track modules, only layers, so this should be `layers`

File: rfcs/20190509-keras-saved-model.md [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+- `metrics`: List of all metric layers attached to this layer and sublayers.
+- `_variables`: List of all variables owned by this object (and not sublayers)
+- `__call__`: Returns the outputs of the call function.
+- `call_and_return_conditional_losses`: Returns the outputs of the call function, as well as a list input-dependent losses (does not include the activity regularizer loss).
+- `activity_regularizer_fn`: Activity regularization function
+- `compile_loss_functions`: List of loss functions added during `model.compile`.
```
Comment (fchollet) [inline rfcs/20190509-keras-saved-model.md]: `compile_losses`

Comment (k-w-w) [reply rfcs/20190509-keras-saved-model.md]: Done

File: rfcs/20190509-keras-saved-model.md (hunk 3 above)
Comment (k-w-w) [reply rfcs/20190509-keras-saved-model.md]: Done

**Comment by karmel** [comment]
# 2019-05-21 Notes from review meeting

*   Any significant changes? No; added a section about compatibility, and it looks like we will be backwards and forwards compatible.
    *   If there is something new added, how will it be compatible? New attrs will be ignored.
    *   As forward compatible as graph def, in other words
    *   Model version numbers? There is a major version in the SM format, and it now stores the TF version
    *   We should increment the SM version number with 2.0 probably.
    *   We had this same problem with graph def, and the solution was that the graph def version is incremented each day.
    *   Integration tests: check with serving, lite, js, hub
*   Do the API changes sound reasonable?
    *   signatures in export: is it possible to always export a default signature even if the user has specified a signature they want to export?
        *   Should be consistent with saved_model.save. Feasible
        *   This would be helpful for serving systems. Always export raw inputs. 
        *   Would this be with serving signature default? Could be private “raw” key.
        *   Conclusion: this can be resolved later, as it would be backwards compat. Should be added to saved_model.save in any case.
    *   What does include_optimizer have to do with whether or not eval fn is included?
        *   If include_optimizer is false, then none of the training objs (loss, metrics, etc) are saved
        *   But eval doesn’t require the optimizer, so it’s a little weird
        *   But if optimizer is not there, loss is not there, so you cannot export eval
        *   Conclusion: export eval_fn if is_compiled=True, irrespective of the val of include_optimizer
        *   Can we change the name of include_optimizer to include_training_function or something? The existing Keras.save/hdf5 already has that arg, so too late to change that, probably.
*   What other aspects of the model should be deserialized/serialized?
    *   Should there be a fn that returns all the losses? call_and_return_total_losses
    *   When you load the model back, you only want the activity regularizers once, whereas the weight regularizers will be called multiple times.
    *   Maybe we could separate out the activity_regularizer losses from the weight regularizers, so that we could call these independently
    *   Imagining that we are loading in w/o keras, is there an easy enough usage pattern? Can we consolidate the places we have to call to get losses, to make these more generically usable outside of keras? regularization_losses and conditional_losses 
    *   If there aren’t existing use-cases, we could add later.
    *   Existing use-case: sonnet wants to follow the same paradigm. Loss structure will be different, but if there is a generic convention, we can follow.
    *   One way to do this would be to have a fn that collects total loss w/o respect for where they come from.
    *   the only way we could do that in SMs is if we had extra variables to track these.
    *   Would that go in tf.Module? Probably not encoded there so much as a convention. A call fn, a regularization fn, and a [one more] fn.
    *   I think it’s important to pull out reg losses versus prediction losses
    *   Separation makes for a fairly complicated API if this is intended to be an API standard.
    *   although end users may never need to know about this if it is a convention, and then authors of these wrappers can work from them.
    *   Conclusion: regularization losses and call_and_return_unconditional_losses would have all losses. We can add a fn that combines these if we want it.
*   Should tf.saved_model.load return a keras object or a generic object?
    *   Cyclic dependence issue if lower level lib returns keras obj
    *   More generally, lower layer should return lower level thing
    *   lower level load should always return the same type of thing, and it would be weird if they got back a different thing depending on what’s in the SM
    *   we can add an additional information message and docs that call out the alternatives for Keras models if the lower level lib sees a keras model
    *   if you try to load in a model saved with the lower level lib, does it load into the keras method?
    *  in theory, if it’s a hub-style saved model we could create a keras-like model. Could be turned in to a subclassed model of sorts
*   is it possible to retrieve the keras json config? It’s saved in the saved model as metadata. Is it easy to retrieve? You would have to load it back in.
*   If you save a tf.module, should this export a similar thing? trainable vars, vars…	
    *   We should centralize on a tf hub style saved model (conventions as discussed above)
    *   But this shouldn’t live with tf.module; we don’t want to specialize module
    *   Logic could live in saved_model.save/load; if a tf.module, make sure to add the hub friendly properties
    *   saved_model.load should always return the same type of thing (standard object), and trainable vars differs between tf.module and keras, so we should make sure it always standardizes… maybe just .variables in that case.

**Comment by k-w-w** [comment]
Thanks for adding the notes, Karmel! 

@ewilderj This can be merged in

**Comment by ewilderj** [comment]
please push a change for the status to "Accepted", and I'll merge. Thank you!

**Comment by k-w-w** [comment]
Done. Thanks!
